sw/device/silicon_creator/rom_ext/rescue.c

```diff
@@ -0,0 +1,220 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/rom_ext/rescue.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/dbg_print.h"
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+#include "sw/device/silicon_creator/lib/xmodem.h"
+
+#include "flash_ctrl_regs.h"
+
+// All of the xmodem functions accept an opaque iohandle pointer.
+// The iohandle is used to facilitate unit tests and doesn't have
+// any function in real firmware.
+#define iohandle NULL
+
+static rescue_state_t rescue_state;
+
+rom_error_t flash_firmware_block(rescue_state_t *state) {
+  if (state->flash_offset == 0) {
+    flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+        .read = kMultiBitBool4True,
+        .write = kMultiBitBool4True,
+        .erase = kMultiBitBool4True,
+    });
+    for (uint32_t addr = state->flash_start; addr < state->flash_limit;
+         addr += FLASH_CTRL_PARAM_BYTES_PER_PAGE) {
+      HARDENED_RETURN_IF_ERROR(
+          flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage));
+    }
+    state->flash_offset = state->flash_start;
+  }
+  if (state->flash_offset < state->flash_limit) {
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_data_write(
+        state->flash_offset, sizeof(state->data) / sizeof(uint32_t),
+        state->data));
+    state->flash_offset += sizeof(state->data);
+  } else {
+    return kErrorRescueImageTooBig;
+  }
+  return kErrorOk;
+}
+
+rom_error_t flash_owner_block(rescue_state_t *state) {
+  // FIXME: validate that we're in a state capable of accepting an owner
+  // block and validate the block before flashing it.
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(&kFlashCtrlInfoPageOwnerSlot1,
+                                                 kFlashCtrlEraseTypePage));
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
+      &kFlashCtrlInfoPageOwnerSlot1, 0, sizeof(state->data) / sizeof(uint32_t),
+      state->data));
+  return kErrorOk;
+}
+
+static void validate_mode(uint32_t mode, rescue_state_t *state) {
+  char *m = (char *)&mode;
+  dbg_printf("\r\nmode: %C%C%C%C\r\n", m[3], m[2], m[1], m[0]);
```

Comment:
are these printf statements going to be part of the production release of rom_ext? If so, we should probably update the function name, and add this to the pentest plan.

Reply:
To be determined: rescue needs some way of sending messages to the user, but it doesn't _need_ to be `printf`.

I'd prefer us to have a robust and secure printing function, but that should be worked out in a separate PR.